### ISSUE TITLE
Use strcred for `twistd mail` authentication options

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4740]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#4740|
|Type|enhancement|
|Created|2010-11-15 00:48:43Z|
|Branch|https://github.com/twisted/twisted/tree/deprecate-mail-passwordfile-4740|

`twistd mail` has a `--passwordfile` option to define credentials for the server.  It should use the strcred mixin to support pluggable authentication systems, instead (with `--passwordfile` kept but deprecated for a while).

Attachments:

* [ticket-4740.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/2b0/2b086baf1e1b6116f287be6b521caec8fc8bc519/bfa6b08da3f374c5da0a4565e4a28a829c938185.patch) (3452 bytes) - added by tpratt on 2011-01-16 22:32:00Z - Add test

<details><summary>Searchable metadata</summary>

```
trac-id__4740 4740
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_deprecate_mail_passwordfile_4740 branches/deprecate-mail-passwordfile-4740
branch_author__cyli cyli
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__easy easy
time__1289782123000000 1289782123000000
changetime__1312567906000000 1312567906000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="tpratt's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|tpratt commented|
|-|-|

I tested this manually and the auth options show up and are accepted.

I need guidance on whether anything here needs to be tested.

**Comment by twisted-trac** [comment]
|<img alt="tpratt's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|tpratt removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_3"></a>|@cyli set owner to tpratt|
|-|-|

Thanks for working on this, tpratt!

 * The warning being emitted is missing the version of Twisted where this deprecation will occur.  If you look at the Twisted Compatibility Policy's ["How to deprecate APIs" section](http://twistedmatrix.com/trac/wiki/CompatibilityPolicy#HowtodeprecateAPIs), there are various methods of deprecating methods and modules, etc. 

 I think the documentation may not be entirely up to date.  For example, you get get a warning string with ```twisted.python.deprecate.getDeprecationWarningString```, although that method doesn't give you an option for a secondary message, like what the method will be deprecated in favor of.  But still, it can be used to give you a good beginning chunk of the deprecation warning.  (see [twisted.python.deprecate](http://twistedmatrix.com/trac/browser/trunk/twisted/python/deprecate.py) and [twisted.python.versions](http://twistedmatrix.com/trac/browser/trunk/twisted/python/versions.py)).

 * There is no unit test that makes sure a deprecation warning is indeed emitted.  The [compatibility policy](http://twistedmatrix.com/trac/wiki/CompatibilityPolicy#HowtodeprecateAPIs) mentions [twisted.trial.unittest.TestCase.flushWarnings](http://twistedmatrix.com/trac/browser/trunk/twisted/trial/unittest.py#L903) as the preferred way to test for deprecation warnings. One example of how to use ```flushWarnings``` is [twisted.web.test.test_web.test_checkExpiredDeprecated](http://twistedmatrix.com/trac/browser/trunk/twisted/web/test/test_web.py#L365). 

 An example of how to test options parsing is given in ```twisted/mail/tap.py```'s test case:  [twisted.mail.test.test_options](http://twistedmatrix.com/trac/browser/trunk/twisted/mail/test/test_options.py).

**Comment by twisted-trac** [comment]
|<img alt="tpratt's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|tpratt commented|
|-|-|

I have etwo issues.

1. None of the options in the t.p.deprecate module seem quite appropriate for a twistd command line option.  I don't see any references to deprecate in any twisted/*/tap.py files.

2. I wrote a test that uses flushWarnings() but it doesn't seem to work if I specify either the test function or the parseOptions() function.  I thought it might be that I'm not passing the correct stacklevel to warnings.warn() but I played with that and it didn't help.  If I don't specify any function to flushWarnings, then it works fine.  I don't understand what's going on here so I don't know the right thing to do.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

Replying to [tpratt](#note_4):
> I have etwo issues.
> 
> 1. None of the options in the t.p.deprecate module seem quite appropriate for a twistd command line option.  I don't see any references to deprecate in any twisted/*/tap.py files.

This is a matter of opinion (our approach to deprecation is, sadly, still evolving) but I think that the `getDeprecationWarningString` function is still appropriate for options.  Perhaps `usage.Options` could grow some specific support for making the interaction of deprecation messages and specific options - or for that matter exception tracebacks and specific options / subcommands - clearer.

> 2. I wrote a test that uses flushWarnings() but it doesn't seem to work if I specify either the test function or the parseOptions() function.  I thought it might be that I'm not passing the correct stacklevel to warnings.warn() but I played with that and it didn't help.  If I don't specify any function to flushWarnings, then it works fine.  I don't understand what's going on here so I don't know the right thing to do.

Please attach this so we can help you diagnose it.

**Comment by twisted-trac** [comment]
|<img alt="tpratt's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|tpratt removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_7"></a>|@cyli commented|
|-|-|

(In [30482]) Branching to 'deprecate-mail-passwordfile-4740'

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_8"></a>|@cyli commented|
|-|-|

Looks good to merge - all tests pass.  I'm just going to remove the unused 'sys' import.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_9"></a>|@cyli commented|
|-|-|

Also renaming the topfile to a .removal instead of a .feature.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_10"></a>|@cyli commented|
|-|-|

(In [30485]) Applied patch, removed unused import, and renamed the topfile to .removal rather than .feature

refs [#4740](https://github.com/twisted/twisted/issues/4740)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_11"></a>|@cyli commented|
|-|-|

(In [30486]) Forgot to commit the news file.  And also, .feature was the correct.

refs [#4740](https://github.com/twisted/twisted/issues/4740)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_12"></a>|@cyli commented|
|-|-|

(In [30487]) Reverting a change that was supposed to be committed to a branch.

refs [#4740](https://github.com/twisted/twisted/issues/4740)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_13"></a>|@cyli commented|
|-|-|

(In [30488]) Rename news file to .feature

refs [#4740](https://github.com/twisted/twisted/issues/4740)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_14"></a>|@cyli set status to `closed`|
|-|-|

(In [30489]) Merge deprecate-mail-passwordfile-4740: strcred for twistd mail authentication

Author: tpratt
Reviewer: cyli
Fixes: [#4740](https://github.com/twisted/twisted/issues/4740)

Use strcred mixin to support pluggable authentication systems for `twistd mail` and deprecated --passwordfile.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

It looks like this didn't actually enable strcred for _twistd mail_, it only added the command line options; the resulting checkers aren't used for anything.  Since this was released already, I filed [#5219](https://github.com/twisted/twisted/issues/5219).
